hal/src/common/gpio/v2.rs

```diff
@@ -0,0 +1,1701 @@
+    /// Pin number
+    const NUM: u8;
+}
+
+/// Value-level `enum` for pin groups
+#[derive(PartialEq, Clone, Copy)]
+pub enum GroupV {
+    A,
+    #[cfg(any(feature = "samd21", feature = "min-samd51g"))]
+    B,
+    #[cfg(feature = "min-samd51n")]
+    C,
+    #[cfg(feature = "min-samd51p")]
+    D,
+}
+
+/// Value-level `struct` representing pin IDs
+#[derive(PartialEq, Clone, Copy)]
+pub struct PinIdV {
+    pub group: GroupV,
+    pub num: u8,
+}
+
+macro_rules! pin_id {
+    ($Group:ident, $Id:ident, $NUM:literal) => {
+        paste! {
+            #[doc = "Pin ID representing pin " $Id]
+            pub enum $Id {}
+            impl Sealed for $Id {}
+            impl PinId for $Id {
+                type Group = [<Group $Group>];
+                const NUM: u8 = $NUM;
+                const ID: PinIdV = PinIdV {
+                    group: GroupV::$Group,
+                    num: $NUM,
+                };
+            }
+        }
+    };
+}
+
+//==============================================================================
+//  Pin trait
+//==============================================================================
+
+/// Type-level meta-type representing pins
+///
+/// All instances of `PinT` implement this trait. When used as a trait bound, it
+/// acts to encapsulate a `PinT`. Without this trait, a completely generic
+/// `PinT` would require two type parameters. When using this trait as a bound,
+/// only one type parameter is required, yet you can still recover each type
+/// parameter of the corresponding `PinT` through the associated types.
+pub trait Pin: Sealed {
+    /// `PinId` of the corresponding `PinT`
+    type Id: PinId;
+    /// `PinMode` of the corresponding `PinT`
+    type Mode: PinMode;
+}
+
+impl<I, M> Sealed for PinT<I, M>
+where
+    I: PinId,
+    M: PinMode,
+{
+}
+
+impl<I, M> Pin for PinT<I, M>
+where
+    I: PinId,
+    M: PinMode,
+{
+    type Id = I;
+    type Mode = M;
+}
+
+//==============================================================================
+//  Optional pins
+//==============================================================================
+
+/// Type-level meta-type representing an optional `PinT`. This trait is
+/// implemented for every `PinT`, as well as for `NoneT`.
+pub trait OptionalPin: Sealed {}
+impl OptionalPin for NoneT {}
+impl<P: Pin> OptionalPin for P {}
+
+/// Type-level meta-type representing a valid `PinT`. When used as a bound, this
+/// trait allows you to reject `NoneT` in favor of `PinT`.
+pub trait SomePin: OptionalPin + Pin {}
+impl<P: OptionalPin + Pin> SomePin for P {}
+
+//==============================================================================
+//  Pin reading & writing
+//==============================================================================
+
+#[inline]
+fn read_pin(group: &GROUP, num: u8) -> bool {
+    let mask: u32 = 1 << num;
+    (group.in_.read().bits() & mask) != 0
+}
+
+#[inline]
+fn write_pin(group: &GROUP, num: u8, bit: bool) {
```

Comment:
So I believe they are completely safe from a memory safety perspective. The `OUTSET/OUTCLR/OUTTGL` registers make these functions completely atomic.

There's a possibility for logic errors if you set the output in the wrong mode, but I don't think logic errors usually qualify for `unsafe`. They are also private, so only module authors need to understand their use.

Reply:
Yeah, it seems like at least the write functions would be memory-safe. I'm not totally sure about the reads; I guess it depends on the hardware semantics. Reading uninitialized memory is undefined behavior, but this is a volatile read of a memory-mapped register, so that rule might not apply here.